python/ray/serve/controller.py

```diff
@@ -835,6 +907,26 @@ def record_multiplexed_replica_info(self, info: MultiplexedReplicaInfo):
         """
         self.deployment_state_manager.record_multiplexed_replica_info(info)
 
+    def set_shutting_down_flag(self):
+        """Set the shutting down flag on controller to signal shutdown in
+        run_control_loop().
+
+        This is used to signal to the controller that it should proceed with shutdown
+        process so it can shutdown gracefully.
+        """
+        self._shutting_down = True
+
+    async def await_controller_shutdown(self):
+        """Await controller to shut down.
+
+        This is used to wait for the controller to finish shutting down and signal to
+        the client that it's safe to exit.
+        """
+        if not self._shutdown.is_set():
+            await self._shutdown.wait()
+
+        return True
```

Comment:
re: my suggestion in `client.py`, you can just combine these into one method:
```suggestion
    async def shutdown(self):
        self._shutting_down = True
        await self._shutdown.wait()
```

Reply:
this is idempotent so no need for two APIs. if we every want to *not* wait for shutdown, we can add `wait: bool = True` kwarg